src/main/kotlin/homeworks/homework3/AvlTree.kt

```diff
@@ -0,0 +1,170 @@
+                val rightChild = node.rightChild
+                if (leftChild == null && rightChild == null) {
+                    size--
+                    return null
+                }
+                if (rightChild == null) {
+                    leftChild?.parent = parentNode
+                    size--
+                    return leftChild
+                }
+                val minimumInRightTree = rightChild.minimum()
+                minimumInRightTree.rightChild = rightChild.removeMinimum()
+                minimumInRightTree.rightChild?.parent = minimumInRightTree
+                minimumInRightTree.leftChild = leftChild
+                leftChild?.parent = minimumInRightTree
+                minimumInRightTree.parent = parentNode
+                size--
+                return minimumInRightTree.balance()
+            }
+        }
+
+        return node.balance()
+    }
+
+    override fun remove(key: K): V? {
+        val value = get(key)
+        root = removeNode(root, key, root?.parent)
+        return value
+    }
+
+    /*
+    In-order iterator for AVL tree. It's private because kotlin stdlib iterator for map is based
+    on entries field and having two public iterators confuses compiler and IDE.
+     */
+    private fun iterator(): Iterator<AvlNode<K, V>> =
+        object : Iterator<AvlNode<K, V>> {
+            private var current: AvlNode<K, V>? = root
+            private var visitedNodes = 0
+
+            override fun hasNext(): Boolean = visitedNodes < size
+
+            @Suppress("ReturnCount")
+            override fun next(): AvlNode<K, V> {
+                if (!hasNext()) {
+                    throw NoSuchElementException("Iterator out of range.")
+                }
+
+                if (visitedNodes == 0) {
+                    current = current?.minimum()
+                    visitedNodes++
+                    return current ?: throw IllegalStateException("Iterator error.")
+                }
+
+                if (current?.rightChild != null) {
+                    current = current?.rightChild?.minimum()
+                    visitedNodes++
+                    return current ?: throw IllegalStateException("Iterator error.")
+                }
+
+                while (current?.parent != null && current?.parent?.leftChild != current) {
+                    current = current?.parent
+                }
+
+                current = current?.parent
+                visitedNodes++
+                return current ?: throw IllegalStateException("Iterator error.")
+            }
+        }
+
+    override fun toString(): String {
+        val sb = StringBuilder()
+        sb.append("{")
+        var i = 0
+        iterator().forEach {
+            if (i > 0) sb.append(", ")
+            sb.append(it.toString())
+            i++
+        }
+        sb.append("}")
+        return sb.toString()
+    }
+
+    fun prettyPrint() {
+        if (root == null) return
+        root?.prettyPrint()
+    }
```

Comment:
`toString` все-таки обязан возвращать `String`, да и хочется соблюдать конвенцию по тому, как отдавать `toString` для Map. В то же время `prettyPrint` сразу пишет в stdout и основан на рекурсии, что собственно и позволяет показать структуру дерева. 
Если переносить `prettyPrint` в `toString`, то надо либо нагло обманывать компилятор, возвращая `""`, либо вносить дополнительную сложность собирая вывод через pre-order обход, при этом надо правильно обсчитывать высоту ноды в дереве.
Кажется слишком сложным для простого принта, хотя именно из-за этой функции превышается лимит функций в классе)

Reply:
По поводу конвенции: да, ты прав, но у нас всё же учебное задание, так что можно ею и пренебречь))) Мне просто не нравится, что в данном классе есть целых две функции, которые занимаются печатью. Как мне кажется этого не должно быть, потому что словарь не должен заниматься печатью) 

По поводу stdout и рекурсии: а что мешает рекурсивно создавать строку?) Вот я тут что-то быстро накидал и вроде всё работает (только над неймингами надо поработать):
```kotlin
internal fun prettyPrint(level: Int = 0): String {
    /* These characters are written as unicode hexadecimals because they
     can easily be confused to their analogs which will not produce ligatures in terminals */
    val append = "\u251C" + "\u2500".repeat(level)

    val stringBuilder = StringBuilder()
    stringBuilder.append("$append$key: $value\n")

    leftChild?.prettyPrint(level + 1)?.let { stringBuilder.append(it) }
    rightChild?.prettyPrint(level + 1)?.let { stringBuilder.append(it) }

    return stringBuilder.toString()
}
```
![image](https://user-images.githubusercontent.com/55441714/163850071-ea5cfffb-7f99-4f3f-b86e-9ee07bcaa16a.png)

Reply:
Ой, мысль о возврате как-то не пришла даже в голову :) Спасибо за помощь